### PR TITLE
[ROCm] Support for gpt2-based model inferencing

### PR DESCRIPTION
**Pull request by ytaous**
When inferencing real gpt2-based model, found some gaps between CUDA and ROCm codebase.

The fixes include:

1. minimum code change to fix tensor shape on Attention Op
2. Support optional output tensor with SkipLayerNorm
3. fix a build error found on MI200

### CODE REVIEW

File: onnxruntime/contrib_ops/rocm/bert/skip_layer_norm_tunable_op.h [hunk 1]
```diff
@@ -48,11 +49,11 @@ Status SkipLayerNormSmallOp(const SkipLayerNormParams<T>* params) {
       !((params->ld <= 1024 && params->ld % VecSize == 0 &&
          params->ld <= ThreadsPerBlock * VecSize && params->ld > (ThreadsPerBlock - GPU_WARP_SIZE) * VecSize)));
   SkipLayerNormKernelSmall<T, ThreadsPerBlock, VecSize><<<dim3(CeilDiv(params->element_count, params->ld)),
-                                                          dim3(ThreadsPerBlock),
-                                                          0, params->stream>>>(
+                                                                      dim3(ThreadsPerBlock),
```
Comment (ytaous) [inline onnxruntime/contrib_ops/rocm/bert/skip_layer_norm_tunable_op.h:52]: I'll fix the indent

File: onnxruntime/python/tools/kernel_explorer/kernels/skip_layer_norm_test.py [hunk 2]
```diff
@@ -64,7 +65,19 @@ def run_skip_layer_norm(batch_size: int, seq_len: int, hidden_size: int, dtype:
     beta_d = ke.DeviceArray(beta)
     y_d = ke.DeviceArray(output_y)
     f = getattr(ke, func)
-    my_op = f(y_d, input_d, skip_d, gamma_d, beta_d, bias_d, epsilon, hidden_size, batch_size * seq_len * hidden_size)
+
+    my_op = f(
+        y_d,
+        output_optional,
```
Comment (PeixuanZuo) [inline onnxruntime/python/tools/kernel_explorer/kernels/skip_layer_norm_test.py:71]: Could you please add a comment here to let us know we are not test `output_optional` like `output_y `?

Comment (ytaous) [reply onnxruntime/python/tools/kernel_explorer/kernels/skip_layer_norm_test.py:71]: will do in my next commit tomorrow, thx for the quick review

Comment (PeixuanZuo) [reply onnxruntime/python/tools/kernel_explorer/kernels/skip_layer_norm_test.py:71]: May encounter some error if there is no `output_optional_d = ke.DeviceArray(output_optional).` and use `output_optional ` directly.

Comment (ytaous) [reply onnxruntime/python/tools/kernel_explorer/kernels/skip_layer_norm_test.py:74]: good catch, will fix it, thx

**Review by PeixuanZuo — APPROVED** [review]

**Pull request merged**